### ISSUE TITLE
Make ready for grocy 3.1.0

### ISSUE THREAD

**Issue by MrJohnZoidberg** [patzly/grocy-android#431]

**Comment by danukefl** [comment]
Adding a note here that the current version of the app connected to a 3.1.0 server will not populate a product from a barcode. This is the only feature that I use that I have found that is not compatible.

**Comment by MrJohnZoidberg** [comment]
@danukefl Can you explain what you mean with "populate a product from a barcode", please?

**Comment by danukefl** [comment]
@MrJohnZoidberg In the Consume, Purchase, Inventory, and Transfer screens, when a valid barcode is scanned, it states that the product is not found. The same barcode can be scanned within the web browser and it will populate properly. It appears that the barcode lookup fails.

This is specifically on server version 3.1.0 with the current beta app targeted for 3.0.1.

**Comment by politas** [comment]
I I find it surprising that the barcode scanning with my phone works for seeming every function except creating a new product. 

On the Create Product screen, It puts up a message saying "This will be implemented soon. Please use your server interface for now. 

Yet barcode scanning works as expected on the Consume, Purchase, Transfer and Inventory pages all allow scanning and entry of all the options

**Comment by MrJohnZoidberg** [comment]
@politas (offtopic) Yes, that's #391. For now, if you scan a barcode on any transaction page (purchase/consume/inventory/transfer) it is written into a small chip below the product field. After that you can enter either an existing product (form will be populated with product details) or a new product name (there will be a popup where you can choose to create a new product with that name) in the product field. After you've created the new product or selected the existing one the barcode chip is still there. This means that the barcode will be linked to the product during the (purchase/consume/inventory/transfer) transaction.
I hope this helps you.

**Comment by MrJohnZoidberg** [comment]
I hope that I have time this weekend to work on the app, sorry for the development pause.

**Comment by MrJohnZoidberg** [comment]
I'm sorry, I won't make the app compatible with server v3.1.0, because that version has two bugs which make it impossible to work with barcodes through the API. The second bug unfortunately makes a workaround for the first bug impossible. More infos here: #433 

So we will have to wait until the next server version, hopefully without one of these bugs or without both of them.

**Comment by MrJohnZoidberg** [comment]
### Simple workaround for those who already have server version 3.1.0 and who have access to the server file system:

This workaround won't corrupt your database and won't have any impact on the website or app, except fixing the bug.
You can easily undo this change if you want to have the bug again.

1. Go to your grocy web directory. Example (your folder structure may be different):
```bash
cd /var/www/grocy
```
2. Open the following file in a text editor (I use Nano for that; you may use `sudo` in front of the command if you have a different user than the web directory):
```bash
sudo nano controllers/BaseApiController.php
```
3. Scroll to `line 24` or in Nano search with `Ctrl`+`W` for `JSON_NUMERIC_CHECK`. You will see this line:
```php
                $response->getBody()->write(json_encode($data, JSON_NUMERIC_CHECK));
```
4. Remove `JSON_NUMERIC_CHECK` together with the comma from the line, so you will get this:
```php
                $response->getBody()->write(json_encode($data));
```
5. Save and close the file (in Nano use `Ctrl`+`O` then `Enter` then `Ctrl`+`X`). Maybe you can also restart your webserver (for me it is `sudo systemctl restart nginx`) but I don't know whether it's necessary.
6. On the purchase screen in our Android app (or any other transaction screen), swipe down with your finger (only one time) to force update of the local data. After that, barcode scans should work again and your product should be recognized.

[Here](https://github.com/grocy/grocy/blob/edfa404ed6b5da4609eee6a664fb5ab54c0e5aa2/controllers/BaseApiController.php#L24) is a link to the line in the grocy repository if you want too look at it.

**Comment by politas** [comment]
I'm actually finding that barcode scanning in the app is working fine with my 3.1.0 server, so I don't understand the issue. What's Barcode Buddy? The only place in the app where I can't scan a barcode is on the Create product screen, where the Barcodes section is greyed out and says "This will be implemented soon. Please use your server interface for now".

**Comment by MrJohnZoidberg** [comment]
Beta 3 is available with server v3.1.1 support and a new barcode scanner.
